File: server/src/routes/agent.ts
import { Router, Request, Response } from 'express';
import fs from 'fs';
import path from 'path';
import { v4 as uuidv4 } from 'uuid';
import { serverModel, CreateServerSchema } from '../models/server.js';
import { deployAgent, getAgentBinaryPath } from '../services/agentDeployer.js';
import { sendToAgent, isAgentConnected } from '../websocket/agentRegistry.js';
import { broadcast } from '../websocket/server.js';
import { config } from '../config.js';
import { logger } from '../utils/logger.js';

const router = Router();

/**
 * POST /api/agent/deploy
 * Deploy the agent to a remote server, creating the server record.
 */
router.post('/deploy', async (req: Request, res: Response) => {
  const parsed = CreateServerSchema.safeParse(req.body);
  if (!parsed.success) return res.status(400).json({ error: parsed.error.flatten() });

  const { password, ...serverData } = parsed.data;

  // Create server record first (so we have an ID for the agent config)
  const server = serverModel.create({ ...serverData, ssh_user: serverData.ssh_user });

  try {
    await deployAgent({
      hostname: serverData.hostname,
      port: serverData.port,
      username: serverData.ssh_user,
      password,
      serverId: server.id,
      serverPort: config.port,
    });

    // The agent should have connected by now (or will soon)
    const updated = serverModel.findById(server.id)!;
    broadcast('server:updated', { server: updated });
    res.status(201).json(updated);
  } catch (err) {
    const errorMsg = err instanceof Error ? err.message : 'Deployment failed';
    logger.error({ err, hostname: serverData.hostname }, 'Agent deployment failed');

    // Delete the server record — no point keeping a failed deployment
    serverModel.delete(server.id);

    res.status(422).json({ error: errorMsg });
  }
});

/**
 * GET /api/agent/binary
 * Serve the compiled agent binary for download.
 */
router.get('/binary', (_req: Request, res: Response) => {
  const binaryPath = getAgentBinaryPath();

  if (!fs.existsSync(binaryPath)) {
    return res.status(404).json({ error: 'Agent binary not found. Build with: cd backup-agent && cargo build --release' });
  }

  res.download(binaryPath, 'backup-agent');
});

/**
 * POST /api/agent/update/:serverId
 * Send an update command to a connected agent.
 */
router.post('/update/:serverId', (req: Request, res: Response) => {
  const { serverId } = req.params;

  const server = serverModel.findById(serverId);
  if (!server) return res.status(404).json({ error: 'Server not found' });

  if (!isAgentConnected(serverId)) {
    return res.status(409).json({ error: 'Agent is not connected' });
  }

  // Send just the path — the agent will prepend its configured server URL
  const sent = sendToAgent(serverId, {
    type: 'agent:update',
    payload: {
      download_path: '/api/agent/binary',
      version: 'latest',
    },
  });

  if (!sent) {
    return res.status(500).json({ error: 'Failed to send update command' });
  }

  // Mark as updating
  serverModel.update(serverId, { agent_status: 'updating' } as any);
  broadcast('server:updated', { server: serverModel.findById(serverId) });

  res.json({ status: 'update_initiated' });
});

/**
 * GET /api/agent/status/:serverId
 * Get agent connection status for a server.
 */
router.get('/status/:serverId', (req: Request, res: Response) => {
  const server = serverModel.findById(req.params.serverId);
  if (!server) return res.status(404).json({ error: 'Server not found' });

  res.json({
    connected: isAgentConnected(req.params.serverId),
    agent_status: (server as any).agent_status,
    agent_version: (server as any).agent_version,
    agent_last_seen: (server as any).agent_last_seen,
  });
});

export default router;

File: backup-server-rs/src/routes/agent.rs
use crate::error::AppError;
use crate::models::server;
use crate::services::agent_deployer;
use crate::state::AppState;
use axum::body::Body;
use axum::extract::{Path, State};
use axum::http::header;
use axum::response::IntoResponse;
use axum::routing::{get, post};
use axum::{Json, Router};
use std::sync::Arc;
use tokio_util::io::ReaderStream;

pub fn router(state: Arc<AppState>) -> Router<Arc<AppState>> {
    Router::new()
        .route("/deploy", post(deploy_agent))
        .route("/binary", get(get_binary))
        .route("/update/{server_id}", post(update_agent))
        .route("/status/{server_id}", get(get_status))
}

async fn deploy_agent(
    State(state): State<Arc<AppState>>,
    Json(body): Json<server::CreateServerRequest>,
) -> Result<(axum::http::StatusCode, Json<server::Server>), AppError> {
    if body.name.is_empty() || body.hostname.is_empty() {
        return Err(AppError::BadRequest("name and hostname are required".into()));
    }

    let password = body.password.clone().unwrap_or_default();
    if password.is_empty() {
        return Err(AppError::BadRequest("password is required for deployment".into()));
    }

    // Create server record first
    let db = state.db.clone();
    let srv = tokio::task::spawn_blocking({
        let body = server::CreateServerRequest {
            name: body.name.clone(),
            hostname: body.hostname.clone(),
            port: body.port,
            ssh_user: body.ssh_user.clone(),
            password: None,
        };
        move || {
            let conn = db.get()?;
            server::create(&conn, &body)
        }
    })
    .await
    .map_err(|e| anyhow::anyhow!(e))??;

    let server_id = srv.id.clone();
    let opts = agent_deployer::DeployOptions {
        hostname: body.hostname.clone(),
        port: body.port as u16,
        username: body.ssh_user.clone(),
        password,
        server_id: srv.id.clone(),
        server_port: state.config.port,
        backup_server_ip: state.config.backup_server_ip.clone(),
    };

    match agent_deployer::deploy_agent(opts, state.agents.clone()).await {
        Ok(()) => {
            let db = state.db.clone();
            let sid = server_id.clone();
            let updated = tokio::task::spawn_blocking(move || {
                let conn = db.get()?;
                server::find_by_id(&conn, &sid)
            })
            .await
            .map_err(|e| anyhow::anyhow!(e))??
            .ok_or_else(|| AppError::Internal(anyhow::anyhow!("Server disappeared")))?;

            state.ui.broadcast("server:updated", serde_json::json!({ "server": updated }));
            Ok((axum::http::StatusCode::CREATED, Json(updated)))
        }
        Err(e) => {
            let error_msg = e.to_string();
            tracing::error!(hostname = %body.hostname, error = %error_msg, "Agent deployment failed");

            // Delete the failed server record
            let db = state.db.clone();
            let sid = server_id;
            let _ = tokio::task::spawn_blocking(move || {
                let conn = db.get()?;
                server::delete(&conn, &sid)
            })
            .await;

            Err(AppError::Unprocessable(error_msg))
        }
    }
}

async fn get_binary() -> Result<impl IntoResponse, AppError> {
    let binary_path = agent_deployer::get_agent_binary_path();

    if !binary_path.exists() {
        return Err(AppError::NotFound(
            "Agent binary not found. Build with: cd backup-agent && cargo build --release".into(),
        ));
    }

    let file = tokio::fs::File::open(&binary_path)
        .await
        .map_err(|e| AppError::Internal(anyhow::anyhow!("Failed to open binary: {}", e)))?;

    let stream = ReaderStream::new(file);
    let body = Body::from_stream(stream);

    Ok((
        [
            (header::CONTENT_TYPE, "application/octet-stream"),
            (
                header::CONTENT_DISPOSITION,
                "attachment; filename=\"backup-agent\"",
            ),
        ],
        body,
    ))
}

async fn update_agent(
    State(state): State<Arc<AppState>>,
    Path(server_id): Path<String>,
) -> Result<Json<serde_json::Value>, AppError> {
    let db = state.db.clone();
    let sid = server_id.clone();
    let srv = tokio::task::spawn_blocking(move || {
        let conn = db.get()?;
        server::find_by_id(&conn, &sid)
    })
    .await
    .map_err(|e| anyhow::anyhow!(e))??;

    if srv.is_none() {
        return Err(AppError::NotFound("Server not found".into()));
    }

    if !state.agents.is_connected(&server_id) {
        return Err(AppError::Conflict("Agent is not connected".into()));
    }

    let sent = state.agents.send_to_agent(
        &server_id,
        serde_json::json!({
            "type": "agent:update",
            "payload": {
                "download_path": "/api/agent/binary",
                "version": "latest",
            },
        }),
    );

    if !sent {
        return Err(AppError::Internal(anyhow::anyhow!("Failed to send update command")));
    }

    // Mark as updating
    let db = state.db.clone();
    let sid = server_id.clone();
    let _ = tokio::task::spawn_blocking(move || {
        let conn = db.get()?;
        server::update_fields(&conn, &sid, &[
            ("agent_status", &"updating" as &dyn rusqlite::types::ToSql),
        ])
    })
    .await;

    Ok(Json(serde_json::json!({ "status": "update_initiated" })))
}

async fn get_status(
    State(state): State<Arc<AppState>>,
    Path(server_id): Path<String>,
) -> Result<Json<serde_json::Value>, AppError> {
    let connected = state.agents.is_connected(&server_id);
    let db = state.db.clone();
    let sid = server_id.clone();
    let srv = tokio::task::spawn_blocking(move || {
        let conn = db.get()?;
        server::find_by_id(&conn, &sid)
    })
    .await
    .map_err(|e| anyhow::anyhow!(e))??;

    match srv {
        Some(s) => Ok(Json(serde_json::json!({
            "connected": connected,
            "agent_status": s.agent_status,
            "agent_version": s.agent_version,
            "agent_last_seen": s.agent_last_seen,
        }))),
        None => Err(AppError::NotFound("Server not found".into())),
    }
}
